File: src/main/java/com/grudus/adventofcode/Day04.java
package com.grudus.adventofcode;

import io.vavr.collection.List;
import io.vavr.collection.Map;

import java.util.Objects;
import java.util.regex.Pattern;

import static java.lang.Integer.parseInt;

public class Day04 extends Day<Integer> {

    @Override
    public Integer firstStar(List<String> input) {
        return List.of(input.mkString("\n").split("\n\n"))
                .map(x -> x.split("\\s+"))
                .map(Passport::fromTokens)
                .filter(Objects::nonNull)
                .size();
    }

    @Override
    public Integer secondStar(List<String> input) {
        List<Passport> validPassports = List.of(input.mkString("\n").split("\n\n"))
                .map(x -> x.split("\\s+"))
                .map(Passport::fromTokens)
                .filter(Objects::nonNull);

        return validPassports.filter(Passport::hasValidValues).size();
    }


    record Passport(Integer byr, Integer iyr, Integer eyr, String hgt, String hcl, String ecl, String pid,
                    String cid) {
        public static Passport fromTokens(String[] tokens) {
            Map<String, String> values = List.of(tokens)
                    .map(token -> token.split(":"))
                    .toMap(token -> token[0], token -> token[1]);

            boolean isValid = List.of("byr", "iyr", "eyr", "hgt", "hcl", "ecl", "pid")
                    .count(values::containsKey) == 7;

            if (!isValid) {
                return null;
            }

            return new Passport(
                    parseInt(values.getOrElse("byr", null)),
                    parseInt(values.getOrElse("iyr", null)),
                    parseInt(values.getOrElse("eyr", null)),
                    values.getOrElse("hgt", null),
                    values.getOrElse("hcl", null),
                    values.getOrElse("ecl", null),
                    values.getOrElse("pid", null),
                    values.getOrElse("cid", null)
            );
        }

        public boolean hasValidValues() {
            return
                    byr >= 1920 && byr <= 2002
                            && iyr >= 2010 && iyr <= 2020
                            && eyr >= 2020 && eyr <= 2030
                            && hasValidHeight()
                    && Pattern.compile("^#(?:[0-9a-f]{6})$").matcher(hcl).matches()
                    && Pattern.compile("amb|blu|brn|gry|grn|hzl|oth").matcher(ecl).matches()
                    && pid.length() == 9;
        }

        private boolean hasValidHeight() {
            final java.util.List<String> groups = Utils.findRegexGroups(Pattern.compile("(\\d+)(cm|in)"), hgt);
            if (groups.size() < 2)
                return false;
            int height = parseInt(groups.get(0));
            String unit = groups.get(1);

            return unit.equals("cm") ? (height >= 150 && height <= 193) : (height >= 59 && height <= 76);
        }
    }
}

File: src/main/python/day04.py
import re

def first_star(passports):
    return len([passport for passport in passports if is_valid_passport(passport) ])

def second_star(passports):
    valid_passports = list(filter(is_valid_passport, passports))
    passport_maps = [create_passport_map(passport.split()) for passport in valid_passports]
    return len(list(filter(has_valid_values, passport_maps)))


def is_valid_passport(passport):
    return has_required_keys(extract_keys(passport.split()))

def extract_keys(passport):
    return [pair.split(":")[0] for pair in passport]

def has_required_keys(password_key):
    return len([key for key in password_key if key != 'cid']) == 7

def create_passport_map(passport):
    return {x.split(":")[0]: x.split(":")[1] for x in passport}

def has_valid_values(passport_map):
    return  (
        (1920 <= int(passport_map['byr']) <= 2002)
    and (2010 <= int(passport_map['iyr']) <= 2020)
    and (2020 <= int(passport_map['eyr']) <= 2030)
    and (has_valid_height(passport_map['hgt']))
    and (bool(re.match(r'^#([0-9a-f]{6})$', passport_map['hcl'])))
    and (passport_map['ecl'] in "amb blu brn gry grn hzl oth".split())
    and (bool(re.match(r'^[0-9]{9}$', passport_map['pid'])))
    )

def has_valid_height(height):
    if height[-2:] == 'cm':
        return 150 <= int(height[:-2]) <= 193
    elif height[-2:] == 'in':
        return 59 <= int(height[:-2]) <= 76
    else:
        return False


if __name__ == "__main__":
    passports = open('src/main/resources/day04/input.txt', 'r').read().split("\n\n")
    print(first_star(passports))
    print(second_star(passports))
